utils/common/src/test/java/org/apache/brooklyn/util/net/ReachabilityCheckerTest.java

```diff
@@ -0,0 +1,89 @@
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.net.HostAndPort;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.mockwebserver.MockResponse;
+import com.google.mockwebserver.MockWebServer;
+import org.apache.brooklyn.test.Asserts;
+import org.apache.brooklyn.util.time.Duration;
+import org.testng.annotations.*;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.concurrent.Executors;
+
+public class ReachabilityCheckerTest {
+    protected MockWebServer server;
+    ListeningExecutorService executor = MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
+
+    @BeforeClass
+    public void setUpClass() throws IOException {
+        executor = MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
+    }
+
+    @BeforeMethod
+    public void setUp() throws IOException {
+        server = new MockWebServer();
+        server.enqueue(new MockResponse().setBody("one message"));
+        server.play();
+    }
+
+    @Test
+    public void testReachibilityChecker() {
+        ReachableSocketFinder finder = new ReachableSocketFinder(connectionChecker(), executor);
+        HostAndPort result = finder.findOpenSocketOnNode(ImmutableList.of(HostAndPort.fromParts(server.getHostName(), server.getPort())), Duration.ONE_MINUTE);
+        Asserts.assertEquals(result, HostAndPort.fromParts(server.getHostName(), server.getPort()));
+    }
+
+    @Test(expectedExceptions = NoSuchElementException.class, expectedExceptionsMessageRegExp = "could not connect to any socket in.*")
+    public void testReachibilityCheckerFails() {
+        ReachableSocketFinder finder = new ReachableSocketFinder(connectionChecker(), executor);
+        finder.findOpenSocketOnNode(ImmutableList.of(HostAndPort.fromParts("localhost", server.getPort() + 1)), Duration.FIVE_SECONDS);
+    }
+
+    @Test
+    public void testReachibilityCheckerChoosesLaunched() {
+        ReachableSocketFinder finder = new ReachableSocketFinder(connectionChecker(), executor);
+        HostAndPort result = finder.findOpenSocketOnNode(ImmutableList.of(HostAndPort.fromParts("localhost", server.getPort() + 1),HostAndPort.fromParts(server.getHostName(), server.getPort())),
+                Duration.FIVE_SECONDS);
+        Asserts.assertEquals(result, HostAndPort.fromParts(server.getHostName(), server.getPort()));
+    }
+
+    protected Predicate<HostAndPort> connectionChecker() {
+        return new Predicate<HostAndPort>() {
+            @Override public boolean apply(HostAndPort input) {
+                return Networking.isReachable(input);
+            }};
+    }
+
+    @AfterMethod
+    public void afterMethod() throws IOException {
+        server.shutdown();
+    }
+
+    @AfterClass
+    public void afterClass() {
+        executor.shutdownNow();
+    }
+}
```

Comment:
You can split the test into two:
 * how `ReachableSocketFinder` behaves - can be used with a mock/recording predicate. Doesn't need network connectivity and relying on specific ports being opened/closed, also quicker to execute.
 * The predicate - this needs network connectivity, but is a very thin wrapper over existing functionality so not too  importnant.